### ISSUE TITLE
SplitTreatments will not re-render when external state changes

### ISSUE THREAD

**Issue by rogerkjoy** [splitio/react-client#42]
The issue is here
https://github.com/splitio/react-client/blob/1ed4a72e49bbddc9325641d7872b7ac5f60df688/src/SplitTreatments.tsx#L23

in the following code...
```
  shouldComponentUpdate(nextProps: Readonly<ISplitTreatmentsProps>) {
    return !shallowEqual(this.props.names, nextProps.names) ||
      !shallowEqual(this.props.attributes, nextProps.attributes);
  }
```

You intentionally do not re-render, even though your component was given a newly constructed render function. The issue this causes is that if that render function is a closure around a local state value, it will not be reflected appropriately...

Here's a quick pseudo-example...
```
function MyThing() {
  const [test, setTest] = useState("not ready");
  setTimeout(() => setTest("ready"), 1000);
  return (
    <SplitTreatments names={features} attributes={attributes}>
      {(response): JSX.Element | null => (<div>{test}</div>)
    </SplitTreatments>
}
```
The rendered div content should change after 1s to ready "ready"; however, due to the above code in shouldComponentUpdate, the child is not re-rendered even though the state has changed.

**Comment by chillaq** [comment]
Hi @rogerkjoy, thanks for the info, checking with the team.

**Comment by arvinio** [comment]
+1 on this issue. `shouldComponentUpdate()` seems to be the culprit [here](https://github.com/splitio/react-client/blob/master/src/SplitTreatments.tsx#L23-L26). We think it's better if the component updates on every change.

**Comment by EmilianoSanchez** [comment]
Hi @arvinio and @rogerkjoy,

You are right. We have released v1.2.3, which includes your suggestion about removing the `shouldComponentUpdate` function to update the component on every change. 

We were using `sCU` to optimize the evaluation of splits and avoid duplicated impressions, but the way it was implemented was leading to the issues you reported. The optimization is done via memoization now.

Please let us know if it works fine on your side. 
Thanks

**Comment by silver-snoopy** [comment]
Hey @EmilianoSanchez,
I am using v1.2.6 and Im experiencing the issue mentioned above.
Upon split update the component is not being rerendered after the splitChanges request is being completed.

Do we have any update on this ticket? 😄

**Comment by michal-filip** [comment]
@silver-snoopy I have just noticed the same issue as you have, but I believe it's different from the one originally described. This issue was about not re-rendering on a non-split state change in the hosting component and seems to be resolved (in 1.2.6 this seems to behave correctly).
What you're describing is re-rendering on a split change, which sounds like a different problem. And it's possible that not everyone would want changes to result in an immediate re-render (it should be an option of `SplitTreatments` and `useTreatments` perhaps?)

**Comment by michal-filip** [comment]
@silver-snoopy It may be ugly and sub-optimal (I'm only just starting with Split), but I ended up building a custom hook that optionally re-renders on split change:

```typescript
import { useClient } from '@splitsoftware/splitio-react';
import { useEffect, useState } from 'react';

import { useMounted } from './useMounted';
import { logger } from '../utils/logger';

export type UseTreatmentOptions = {
  updateOnChange?: boolean;
};

export const useTreatment = (
  name: string,
  { updateOnChange = true }: UseTreatmentOptions = {},
) => {
  const isMounted = useMounted();
  const [isReady, setReady] = useState(false);
  const [treatment, setTreatment] = useState('control');
  const client = useClient();

  useEffect(() => {
    if (client?.ready) {
      client?.ready().then(() => {
        isMounted && setReady(true);
      });
    }
  }, [client?.ready]);

  useEffect(() => {
    if (client) {
      setTreatment(client.getTreatment(name));

      if (updateOnChange) {
        const listener = () => {
          logger.info('treatment updated', client.getTreatment(name));
          setTreatment(client.getTreatment(name));
        };

        client.on(client.Event.SDK_UPDATE, listener);

        return () => {
          client.off(client.Event.SDK_UPDATE, listener);
        };
      }
    }

    return undefined;
  }, [isReady]);

  return treatment;
};
```
It would be nice having something like this polished and baked into the library though.

**Comment by silver-snoopy** [comment]
@michal-filip thanks for the suggestion! 🎉 
The proposed solution looks good. I would also endorse the configurable solution of the useTreatments hooks which would re-render based on given parameter.

**Comment by EmilianoSanchez** [comment]
Hi @silver-snoopy , @michal-filip ,

Sorry for the long delay on answering to the thread. 

As you mention, `SplitTreatments`, `useTreatments` and `useClient` are missing that configurable parameter, equivalent to the `updateOnSdkUpdate` param of `SplitFactory` and `SplitClient` components. For consistency and flexibility that is a very interesting feature to add, but we are not able to prioritize this work at the moment.

In the meantime, there are some potential workarounds you could consider, like the one @michal-filip mentioned.

Another option is to rely on `SplitClient` component:

```
const App = () => {
  return (
    <SplitFactory factory={factory} /* updateOnSdkUpdate is false by default */ >
      <SplitClient updateOnSdkUpdate={true} >
        <CompA />
      </ SplitClient>
      <SplitClient updateOnSdkUpdate={false} >
        <CompB />
      </ SplitClient>
    </SplitFactory>
  );
}
```

Here `SplitClient` is used to overwrite the `updateOnSdkUpdate` param from the SplitFactory (the user key is not changed). 
Therefore `CompA` re-renders on a SDK update but `CompB` doesn't.

**Comment by EmilianoSanchez** [comment]
I am closing this issue with the following update:

We have released React SDK v1.10.0, which includes a rework of hooks: `useClient`, `useTreatments` and `useManager` hooks were deprecated and replaced by `useSplitClient`, `useSplitTreatments` and `useSplitManager` hooks respectively. 

The new hooks return the Split context object, along with the SDK client, treatments, and manager respectively. This enables direct access to status properties like `isReady`, eliminating the need for using the `useContext(SplitContext)` hook or the client's `ready` promise. Additionally, `useSplitClient` and `useSplitTreatments` accept the same "update" options as `SplitFactory` and `SplitClient` components to control when the hooks will re-render: `updateOnSdkReady`, `updateOnSdkReadyFromCache`, `updateOnSdkUpdate`, and `updateOnSdkTimedout`. 

The following code snippet shows how to refactor the usage of the deprecated hooks:

```javascript
// Old hooks
import { useClient, useTreatments, useManager } from '@splitsoftware/splitio-react';

const client = useClient('user-id');
const treatments = useTreatments(['feature_1'], { att1: 'att1' }, 'user-id');
const manager = useManager();

// New hooks
import { useSplitClient, useSplitTreatments, useSplitManager } from '@splitsoftware/splitio-react';

const { client, isReady, ... } = useSplitClient({ splitKey: 'user-id' });
const { treatments, isReady, ... } = useSplitTreatments({ names: ['feature_1'], attributes: { att1: 'att1' }, splitKey: 'user-id' });
const { manager, isReady, ... } = useSplitManager();
```

This way, the code snippet provided [here](https://github.com/splitio/react-client/issues/42#issuecomment-884974784) can be simplified as follows:

```javascript
import { useSplitTreatments } from '@splitsoftware/splitio-react';
import { logger } from '../utils/logger';
import { useEffect } from 'react';

export type UseTreatmentOptions = {
  updateOnChange?: boolean;
};

export const useTreatment = (
  name: string,
  { updateOnChange = true }: UseTreatmentOptions = {},
) => {
  const { treatments, isReady, isReadyFromCache } = useSplitTreatments({
    names: [name],
    updateOnSdkUpdate: updateOnChange, // updateOnSdkUpdate is false by default
    // updateOnSdkReady: true // true by default
    // updateOnSdkReadyFromCache: true // true by default
    // updateOnSdkTimedout: false // false by default
  });

  // treatment is 'control' if `isReady` and `isReadyFromCache` are false, or if the feature flag name does not exist.
  const treatment = treatments[name].treatment;

  useEffect(() => {
    logger.info('treatment updated', treatment);
  }, [treatment]);

  return treatment;
};
```
